lib/modules.nix

```
@@ -412,10 +416,34 @@ rec {
       # modules recursively. It returns the final list of unique-by-key modules
       filterModules = modulesPath: { disabled, modules }:
         let
-          moduleKey = m: if isString m && (builtins.substring 0 1 m != "/")
-            then toString modulesPath + "/" + m
-            else toString m;
-          disabledKeys = map moduleKey disabled;
+          moduleKey = file: m:
+            if isString m
+            then
+              if builtins.substring 0 1 m == "/"
+              then m
+              else toString modulesPath + "/" + m
+
+            else if isConvertibleWithToString m
+            then
+              if m?key
+              then
+                throw "Module `${file}` contains a disabledModules item that is an attribute set that used to be converted to a string for the purpose of retrieving the module key, which is a unique identifier of a module. However it also has a `key` attribute, which makes the intent of the attribute set ambiguous in light of the module system's improved support for disabling attribute-set-based modules. To recover the previous behavior of your module, please apply `toString` to the offending disabledModules item. Future versions of the module system will pick the more explicit `key` attribute instead."
```

```suggestion
                throw "Module `${key}` contains a disabledModules item that is an attribute set that can be converted to a string (${toString m}) but also has a `.key` attribute (${m.key}) with a different value. This makes it ambigous which module should be disabled."
```

Let's change the semantics so that it checks for `isConvertibleWithToString` and `? key`, if only one of them is true use that, if both are true get their values and ensure they are the same, otherwise error as above. And I don't think we need to be super explicit about how to this error then, it's fairly obvious if we print the two differing values.

This way we don't need to worry about deprecating anything or changing behavior. And we can still deprecate allowing `isConvertibleWithToString` if we wanted to.

Also I'm suggesting `file -> key` here for the error messages, because `key`'s don't uniquely identify modules.